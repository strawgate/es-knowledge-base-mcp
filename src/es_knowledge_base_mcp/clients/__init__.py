"""Clients for interacting with external services."""
