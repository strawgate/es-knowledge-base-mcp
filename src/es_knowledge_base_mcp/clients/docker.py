from contextlib import asynccontextmanager
import logging
import io
import tarfile
import datetime
from aiodocker.docker import Docker, DockerContainer
from aiodocker.exceptions import DockerError
from dataclasses import dataclass
from typing import AsyncIterator, List, Dict, Any, Optional
from es_knowledge_base_mcp.models.constants import BASE_LOGGER_NAME

logger = logging.getLogger(BASE_LOGGER_NAME).getChild("utils").getChild("docker")


# region File Injection
@dataclass
class InjectFile:
    """Represents a file to be injected into a container."""

    filename: str  # Full path *inside* the container (e.g., "/app/config.yml")
    content: str  # File content as a string

    def to_tar_stream(self) -> io.BytesIO:
        """Converts the file content to a tar stream."""
        tar_stream = io.BytesIO()

        with tarfile.open(fileobj=tar_stream, mode="w") as tar:
            file_bytes = self.content.encode("utf-8")
            tarinfo = tarfile.TarInfo(name=self.filename.lstrip("/"))
            tarinfo.size = len(file_bytes)
            tarinfo.mtime = int(datetime.datetime.now(datetime.timezone.utc).timestamp())
            tar.addfile(tarinfo, io.BytesIO(file_bytes))

        tar_stream.seek(0)

        return tar_stream


@asynccontextmanager
async def handle_errors(operation: str) -> AsyncIterator[None]:
    """Context manager for error handling."""
    logger.debug(f"Attempting {operation}...")

    try:
        yield
    except DockerError as e:
        logger.error(f"Docker error during {operation}: {e}")
        raise
    except Exception as e:
        logger.error(f"Unexpected error during {operation}: {e}")
        raise

    logger.debug(f"Completed {operation}.")


# region Start Container


async def start_container_with_files(
    docker_client: Docker,
    image_name: str,
    command: List[str],
    files_to_inject: List[InjectFile],
    labels: Dict[str, str],
    container_name: Optional[str] = None,
) -> str:
    """Creates a container, injects files into it, and then starts it.

    Args:
        docker_client: Docker client instance.
        image_name: Name of the Docker image to use.
        command: Command to run in the container.
        files_to_inject: List of files to inject into the container.
        labels: Labels to apply to the container.
        container_name: Optional name for the container.

    Returns:
        str:  12-character container ID.

    Example:
        >>> container_id = await start_container_with_files(
                docker_client,
                image_name="my_image:latest",
                command=["python", "app.py"],
                files_to_inject=[InjectFile(filename="/app/config.yml", content="config content")],
                labels={"env": "test"},
                container_name="my_container"
            )
    12-character container ID is returned.
    """

    container: Optional[DockerContainer] = None

    container_config = {
        "Image": image_name,
        "Cmd": command,
        "Labels": labels or {},
        "HostConfig": {"AutoRemove": False},
        "MemoryReservation": 512 * 1024 * 1024,  # 512 MB
    }

    logger.debug(f"Preparing container '{container_name or 'unnamed'}' for image '{image_name}' with labels {labels}")

    async with handle_errors("container setup"):
        container = await docker_client.containers.create(config=container_config, name=container_name)
        logger.debug(f"Created container '{container.id}'.")

    if files_to_inject:
        logger.debug(f"Preparing to inject {len(files_to_inject)} file(s) into container '{container.id}'.")
        [await container.put_archive(path="/", data=file.to_tar_stream()) for file in files_to_inject]

    async with handle_errors("container start"):
        await container.start()

    return container.id


# endregion Start Container


# region Container Info
async def get_containers(docker_client: Docker, label_filter: str, all_containers: bool = False) -> list[DockerContainer]:
    """Lists containers matching a label filter, returning basic info."""
    label_filter_str = label_filter if "=" in label_filter else f"{label_filter}"

    logger.debug(f"Listing containers with label filter '{label_filter_str}'...")

    async with handle_errors("container list"):
        containers = await docker_client.containers.list(all=all_containers, filters={"label": [label_filter_str]})

    logger.debug(f"Found {len(containers)} container(s) matching label filter '{label_filter_str}'.")

    return containers


async def get_containers_details(docker_client: Docker, label_filter: str, all_containers: bool = False) -> List[Dict[str, Any]]:
    """Lists containers matching a label filter, returning detailed info."""

    containers = await get_containers(docker_client, label_filter, all_containers)

    return [container._container for container in containers]


async def container_logs(docker_client: Docker, container_id: str) -> str:
    """Retrieves logs for a specific container ID."""

    logger.debug(f"Retrieving logs for container '{container_id}'...")

    async with handle_errors("container logs"):
        container = await docker_client.containers.get(container_id)
        logs = await container.log(stdout=True, stderr=True)

    logger.debug(f"Retrieved logs for container '{container_id}'.")

    return "".join(logs)


# endregion Container Info

# region Cleanup Container


async def remove_container(docker_client: Docker, container_id: str) -> None:
    """Removes containers matching a label filter."""
    logging.debug(f"Removing container '{container_id}'...")

    async with handle_errors("container removal"):
        container = await docker_client.containers.get(container_id)
        await container.delete(force=True)

    logger.debug(f"Removed container '{container_id}'.")


async def remove_containers(docker_client: Docker, label_filter: str) -> None:
    """Removes containers matching a label filter."""

    containers = await get_containers(docker_client, label_filter)

    logger.debug(f"Removing {len(containers)} container(s) matching label filter '{label_filter}'...")

    async with handle_errors("container removal"):
        [await container.delete(force=True) for container in containers]

    logger.debug(f"Removed {len(containers)} container(s) matching label filter '{label_filter}'.")


# endregion Cleanup Container
