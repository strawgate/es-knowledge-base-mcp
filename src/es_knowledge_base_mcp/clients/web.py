"""Utility functions for web-related operations."""

from typing import Any
from urllib.parse import urljoin, urlparse, urlunparse

import requests
from bs4 import BeautifulSoup, Tag
from fastmcp.utilities.logging import get_logger

logger = get_logger("knowledge-base-mcp.utils")


async def extract_urls_from_webpage(url: str, domain_filter: str | None = None, path_filter: str | None = None) -> dict[str, Any]:
    """Extracts all unique URLs from a given webpage, stripping fragments and query parameters. Optionally,
    filters URLs based on a specific domain and path. Also extracts meta robots directives.

    Args:
        url: The URL of the webpage to extract URLs from.
        domain_filter: Optional filter to restrict URLs to a specific domain.
        path_filter: Optional filter to restrict URLs to a specific path.

    Returns:
        A dictionary containing page directives and categorized URLs:
        {
            "page_is_noindex": bool,
            "page_is_nofollow": bool, # From meta tag
            "urls_to_crawl": List[str], # Unique, sorted list
            "skipped_urls": List[str]  # Unique, sorted list
        }

    Example:
        >>> result = await extract_urls_from_webpage("http://example.com")
        >>> result["urls_to_crawl"]
        ["http://example.com/page1", "http://example.com/page2"]

    """
    response = requests.get(url, timeout=10)
    response.raise_for_status()
    soup = BeautifulSoup(response.content, "html.parser")

    # Extract Meta Robots Directives
    page_is_noindex = False
    page_is_nofollow = False
    meta_robots = soup.find("meta", attrs={"name": lambda x: x and x.lower() == "robots"})  # type: ignore
    if meta_robots and meta_robots.get("content"):  # type: ignore
        content = meta_robots.get("content", "").lower()  # type: ignore
        if "noindex" in content:
            page_is_noindex = True
        if "nofollow" in content:
            page_is_nofollow = True

    # Modify Link Extraction & Categorization
    urls_to_crawl = set()
    skipped_urls = set()

    for a in soup.find_all("a", href=True):
        if not isinstance(a, Tag) or not a.get("href"):
            continue

        href = a["href"]
        is_nofollow_link = False
        rel_attr = a.get("rel", [])  # type: ignore
        if isinstance(rel_attr, str):
            rel_attr = rel_attr.split()
        if "nofollow" in [rel.lower() for rel in rel_attr]:  # type: ignore
            is_nofollow_link = True

        # Existing logic to build absolute URL (urljoin)
        absolute_url = urljoin(url, str(href))

        # Existing logic to parse and clean URL (urlparse, urlunparse)
        parsed_url = urlparse(absolute_url)
        cleaned_url = urlunparse(parsed_url._replace(fragment="", query=""))
        parsed_domain = parsed_url.scheme + "://" + parsed_url.netloc

        # Existing filter logic (domain_filter, path_filter)
        if path_filter and not parsed_url.path.startswith(path_filter):
            continue
        if domain_filter and parsed_domain != domain_filter:
            continue

        # Add to appropriate set based on nofollow status
        if is_nofollow_link:
            skipped_urls.add(cleaned_url)
        else:
            urls_to_crawl.add(cleaned_url)

    # Update Return Statement
    return {
        "page_is_noindex": page_is_noindex,
        "page_is_nofollow": page_is_nofollow,
        "urls_to_crawl": sorted(urls_to_crawl),
        "skipped_urls": sorted(skipped_urls),
    }
