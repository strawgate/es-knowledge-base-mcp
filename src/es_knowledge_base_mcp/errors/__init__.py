"""Custom exceptions for the knowledge base MCP."""
