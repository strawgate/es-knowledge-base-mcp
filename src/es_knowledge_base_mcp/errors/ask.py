from es_knowledge_base_mcp.errors.server import KnowledgeBaseMCPBaseError


class AskError(KnowledgeBaseMCPBaseError):
    """Raised when the Ask operation encounters an error."""

    msg: str = "Unknown Ask error"


class AskQuestionAnswerError(AskError):
    """Raised when the Ask operation fails to answer a question."""

    msg: str = "Unable to answer the question"
