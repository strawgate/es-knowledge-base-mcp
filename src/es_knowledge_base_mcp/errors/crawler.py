from es_knowledge_base_mcp.errors.server import KnowledgeBaseMCPBaseError


class CrawlerError(KnowledgeBaseMCPBaseError):
    """Raised when the crawler encounters an error."""

    msg: str = "Unknown Crawler error"

    def __init__(self, message: str):
        super().__init__(message)
        self.msg = message


class CrawlerValidationError(CrawlerError):
    """Raised when the crawler encounters an error validating the URL to crawl."""

    msg: str = "Error while validating the URL to crawl"


class CrawlerValidationTooManyURLsError(CrawlerValidationError):
    """Raised when the crawler encounters too many URLs while validating a url to crawl."""

    msg: str = "Too many URLs to crawl"


class CrawlerValidationHTTPError(CrawlerValidationError):
    """Raised when the crawler encounters an error validating the URL to crawl."""

    msg: str = "Error while validating the URL to crawl"


class CrawlerDockerError(CrawlerError):
    """Raised when the crawler encounters a Docker error."""

    msg: str = "Unknown Crawler Docker error"


class CrawlerDockerImageError(CrawlerDockerError):
    """Raised when the crawler encounters an error checking or pulling Docker images."""

    msg: str = "Error checking or pulling the Docker image for the Crawler"


class CrawlerDockerContainerError(CrawlerDockerError):
    """Raised when the crawler encounters a Docker container error."""

    msg: str = "Error starting the Docker container for the Crawler"


class CrawlerValidationNoIndexNofollowError(CrawlerValidationError):
    """Raised when a seed URL is marked with both noindex and nofollow."""
