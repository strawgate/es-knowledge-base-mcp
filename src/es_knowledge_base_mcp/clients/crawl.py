from __future__ import annotations
from contextlib import asynccontextmanager
from uuid import uuid4
import yaml
import urllib.parse
from typing import TYPE_CHECKING, List, Dict, Any
from aiodocker.docker import Docker
from aiodocker.exceptions import DockerError


from es_knowledge_base_mcp.clients import docker as docker_utils
from es_knowledge_base_mcp.clients.docker import InjectFile
from es_knowledge_base_mcp.errors.crawler import (
    CrawlerDockerError,
    CrawlerError,
    CrawlerValidationHTTPError,
    CrawlerValidationTooManyURLsError,
    CrawlerValidationNoIndexNofollowError,
)
from requests import HTTPError

from fastmcp.utilities.logging import get_logger
from es_knowledge_base_mcp.clients.web import extract_urls_from_webpage

from es_knowledge_base_mcp.models.settings import CrawlerSettings

if TYPE_CHECKING:
    from es_knowledge_base_mcp.models.settings import ElasticsearchSettings

logger = get_logger("knowledge-base-mcp.crawl")

# endregion Crawler Settings


class Crawler:
    """Handles the logic for crawling websites using Docker."""

    docker_socket: str | None

    docker_client: Docker

    settings: CrawlerSettings

    # Container Label used to identify containers managed by this component
    MANAGED_BY_LABEL = "managed-by"
    MANAGED_BY_VALUE = "mcp-crawler"

    DOMAIN_LABEL = "crawl-domain"
    CRAWL_CONFIG_PATH = "/config/crawl.yml"

    def __init__(
        self,
        settings: CrawlerSettings,
        elasticsearch_settings: ElasticsearchSettings,
    ):
        """Initializes the Crawler component."""

        self.elasticsearch_settings = elasticsearch_settings

        self.settings = settings
        self.docker_socket = settings.docker_socket

        logger.debug("Crawler component initialized with settings: %s", self.settings)

    async def async_init(self) -> None:
        """Initializes the Docker client for the crawler."""
        logger.debug("Initializing Docker client for Crawler...")

        try:
            self.docker_client = Docker(url=self.docker_socket)
        except DockerError as e:
            logger.error("Failed to initialize Docker client: %s", e)
            raise CrawlerError(f"Failed to initialize Docker client: {e}") from e

        logger.debug("Docker client initialized.")

    async def async_shutdown(self) -> None:
        """Cleans up the Docker client for the crawler."""
        try:
            await self.docker_client.close()
        except Exception as e:
            logger.error("Failed to close Docker client: %s", e)
            raise CrawlerError(f"Failed to close Docker client: {e}") from e

    @asynccontextmanager
    async def handle_errors(self, operation: str = "Docker operation"):
        """Context manager to handle errors during Docker operations."""
        try:
            yield
        except DockerError as e:
            message = f"Docker {operation} failed: {e.message}"
            logger.error(message)
            raise CrawlerDockerError(message) from e
        except Exception as e:
            message = f"Unexpected error during {operation}: {e}"
            logger.error(message)
            raise CrawlerError(message) from e

    # region Prepare Config
    @classmethod
    async def _prepare_crawl_config_file(
        cls,
        domain: str,
        seed_url: str,
        filter_pattern: str,
        elasticsearch_index_name: str,
        crawler_es_settings: dict[str, Any],
        exclude_paths: list[str] | None = None,
    ) -> InjectFile:
        """
        Generates the crawler configuration content (as YAML) in memory. This configuration is the Crawler configuration
        which is documented in https://github.com/elastic/crawler/blob/main/docs/CONFIG.md.

        Returns:
            InjectFile: An object containing the generated config content and target path within the container.
        """

        additional_exclusion_rules = []

        if exclude_paths is not None:
            # if the path is a full url we need to extract the path from it
            trimmed_exclude_paths = [urllib.parse.urlparse(path).path for path in exclude_paths]
            additional_exclusion_rules.extend([{"policy": "deny", "type": "begins", "pattern": path} for path in trimmed_exclude_paths])

        config = {
            "domains": [
                {
                    "url": domain,
                    "seed_urls": [seed_url],
                    "crawl_rules": [
                        *additional_exclusion_rules,
                        {"policy": "allow", "type": "begins", "pattern": filter_pattern},
                        {"policy": "deny", "type": "regex", "pattern": ".*"},
                    ],
                }
            ],
            "log_level": "DEBUG",
            "output_sink": "elasticsearch",
            "output_index": elasticsearch_index_name,
            "elasticsearch": {
                **crawler_es_settings,
            },
        }

        config_container_path = cls.CRAWL_CONFIG_PATH

        return InjectFile(filename=config_container_path, content=yaml.dump(config, indent=2))

    # endregion Prepare Config

    @classmethod
    def derive_crawl_params(cls, url: str) -> Dict[str, str]:
        """Derives crawl parameters using a heuristic based on the URL. Intelligently determine
        the right parameters to use based on the URL provided:
        1. If the url has a file extension, we assume it's a file. We then crawl everything that matches the url
            up to the last `/` character.
        2. If the url doesn't have a file extension, we assume it's a directory. We then crawl everything url
            starts with the url provided.

        Args:
            url: The URL to process.

        Returns:
            A dictionary containing "page_url", "domain", "filter_pattern", and "elasticsearch_index_name".
        """

        parsed = urllib.parse.urlparse(url)

        path = parsed.path
        filter_pattern = path

        # if end of the url is a file, we need to crawl everything that matches the url up to the last `/` character
        if not path.endswith("/") and "." in path.split("/")[-1]:
            filter_pattern = path[: path.rfind("/") + 1] or "/"

        return {
            "seed_url": url,
            "domain": parsed.scheme + "://" + parsed.netloc,
            "filter_pattern": filter_pattern,
        }

    # endregion Crawl Parameters
    # region Crawl Validation

    @classmethod
    async def validate_crawl(cls, url: str, max_child_page_limit: int = 500) -> Dict[str, Any]:
        """
        Validates whether the target URL is suitable for crawling.

        Checks for potential issues such as excessive child URLs or 'noindex'/'nofollow' directives.

        Args:
            url: The URL to validate.
            max_child_page_limit: The maximum allowed number of child pages.

        Returns:
            A dictionary containing derived crawl parameters if validation is successful.

        Raises:
            CrawlerValidationHTTPError: If there's an HTTP error fetching the URL.
            CrawlerValidationNoIndexNofollowError: If the page has both 'noindex' and 'nofollow' directives.
            CrawlerValidationTooManyURLsError: If the number of child URLs exceeds the limit.
        """
        logger.debug(f"Validating url {url} for crawling")

        crawl_parameters = cls.derive_crawl_params(url)

        logger.debug(f"Derived crawl parameters: {crawl_parameters}")

        try:
            extraction_result = await extract_urls_from_webpage(
                url=url, domain_filter=crawl_parameters["domain"], path_filter=crawl_parameters["filter_pattern"]
            )

        except HTTPError as e:
            reason = f"Could not validate crawl. Failed to extract URLs from {url}: {e}"
            logger.error(reason)
            raise CrawlerValidationHTTPError(message=reason)

        if extraction_result["page_is_noindex"] and extraction_result["page_is_nofollow"]:
            reason = f"Validation failed: Seed URL {url} is marked with both 'noindex' and 'nofollow'."
            logger.error(reason)
            raise CrawlerValidationNoIndexNofollowError(message=reason)

        num_urls = len(extraction_result["urls_to_crawl"])
        logger.debug(f"Found {num_urls} URLs to crawl (excluding nofollow links).")

        if num_urls > max_child_page_limit:
            reason = f"Could not validate crawl. Excessive child URLs ({num_urls} > {max_child_page_limit}). Validate that you're crawling a specific enough URL or consider setting max_child_page_limit."
            logger.error(reason)
            raise CrawlerValidationTooManyURLsError(message=reason)

        return crawl_parameters

    # endregion Crawl Validation

    # region Image Handling
    async def pull_crawler_image(self) -> None:
        """Pulls the configured crawler Docker image."""
        image_name = self.settings.docker_image

        logger.debug(f"Pulling Docker image '{image_name}' for crawler...")

        await self.docker_client.images.pull(image_name)

    # endregion Image Handling

    # region Start Crawl

    async def crawl_domain(
        self,
        domain: str,
        seed_url: str,
        filter_pattern: str,
        elasticsearch_index_name: str,
        exclude_paths: list[str] | None = None,
    ) -> str:
        """
        Starts a crawl job asynchronously by launching a container.

        Args:
            domain (str): The domain to crawl.
            seed_url (str): The seed URL to start the crawl from.
            filter_pattern (str): The filter pattern for the crawl.
            elasticsearch_index_name (str): The suffix for the output index name.
            exclude_paths (str | None): Optional paths to exclude from the crawl.

        Returns:
            str: The ID of the started container.

        Example:
            >>> crawl_domain("http://example.com", "http://example.com/start", "/filter", "example_com")
            "container_id_123456"
        """

        logger.debug(f"Attempting to start crawl for domain '{domain}' -> index '{elasticsearch_index_name}'")

        config_file_to_inject = await self._prepare_crawl_config_file(
            domain=domain,
            seed_url=seed_url,
            filter_pattern=filter_pattern,
            exclude_paths=exclude_paths,
            elasticsearch_index_name=elasticsearch_index_name,
            crawler_es_settings=self.elasticsearch_settings.to_crawler_settings(),
        )

        random_id = uuid4().hex[:8]

        async with self.handle_errors("starting crawl container"):
            container_id = await docker_utils.start_container_with_files(
                docker_client=self.docker_client,
                container_name=f"mcp-crawler-{elasticsearch_index_name}-{random_id}",
                image_name=self.settings.docker_image,
                command=["ruby", "bin/crawler", "crawl", config_file_to_inject.filename],
                files_to_inject=[config_file_to_inject],
                labels={
                    self.MANAGED_BY_LABEL: self.MANAGED_BY_VALUE,
                    self.DOMAIN_LABEL: domain,
                },
            )

            logger.info(f"Started crawl for domain '{domain}' -> index '{elasticsearch_index_name}' with container ID '{container_id}'")

            return container_id

    # endregion Crawl

    # region Manage Crawls
    async def list_crawls(self) -> List[Dict[str, Any]]:
        """Lists all containers managed by this crawler component."""
        label_filter = f"{self.MANAGED_BY_LABEL}={self.MANAGED_BY_VALUE}"

        async with self.handle_errors("listing crawl containers"):
            return await docker_utils.get_containers_details(self.docker_client, label_filter)

    async def get_crawl_logs(self, container_id: str) -> str:
        """Gets logs for a specific crawl container.

        Args:
            container_id (str): The ID of the container to retrieve logs for."""
        async with self.handle_errors("getting crawl logs"):
            return await docker_utils.container_logs(self.docker_client, container_id)

    async def stop_crawl(self, container_id: str) -> None:
        """Stops and removes a specific crawl container.

        Args:
            container_id (str): The ID of the container to stop and remove.
        """
        async with self.handle_errors("stopping crawl container"):
            await docker_utils.remove_container(self.docker_client, container_id)

    # endregion Manage Crawls

    # region Cleanup

    async def remove_completed_crawls(self) -> Dict[str, Any]:
        """
        Removes all completed (exited) crawl containers managed by this component.
        Returns a summary of the operation.
        """
        logger.info("Attempting to remove completed crawl containers...")

        async with self.handle_errors("removing completed crawl containers"):
            crawler_containers = await docker_utils.get_containers(
                self.docker_client, f"{self.MANAGED_BY_LABEL}={self.MANAGED_BY_VALUE}", all_containers=True
            )

        completed_crawls = [container for container in crawler_containers if container["State"] == "exited"]

        logger.debug(f"Found {len(completed_crawls)} completed crawls.")

        async with self.handle_errors("removing completed crawl containers"):
            [await docker_utils.remove_container(self.docker_client, container["Id"]) for container in completed_crawls]

        logger.info(f"Removed {len(completed_crawls)} completed crawls.")

        return {
            "removed": len(completed_crawls),
            "total": len(crawler_containers),
        }

    # endregion Cleanup
