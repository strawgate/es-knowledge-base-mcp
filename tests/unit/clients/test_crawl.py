from unittest.mock import AsyncMock, patch

import pytest
from requests import HTTPError

from es_knowledge_base_mcp.clients.crawl import Crawler
from es_knowledge_base_mcp.clients.docker import InjectFile
from es_knowledge_base_mcp.errors.crawler import (
    CrawlerValidationHTTPError,
    CrawlerValidationNoIndexNofollowError,
    CrawlerValidationTooManyURLsError,
)  # Import new error


@pytest.mark.parametrize(
    "url, expected_params",
    [
        (
            "http://example.com/docs/file.html",
            {
                "seed_url": "http://example.com/docs/file.html",
                "domain": "http://example.com",
                "filter_pattern": "/docs/",
            },
        ),
        (
            "https://anothersite.com/guide/",
            {
                "seed_url": "https://anothersite.com/guide/",
                "domain": "https://anothersite.com",
                "filter_pattern": "/guide/",
            },
        ),
        (
            "http://example.com",
            {
                "seed_url": "http://example.com",
                "domain": "http://example.com",
                "filter_pattern": "",
            },
        ),
        (
            "http://example.com/file.txt",
            {
                "seed_url": "http://example.com/file.txt",
                "domain": "http://example.com",
                "filter_pattern": "/",
            },
        ),
    ],
    ids=[
        "URL with file extension",
        "URL with trailing slash",
        "Root URL",
        "Root URL with file extension",
    ],
)
def test_derive_crawl_params(url, expected_params):
    """Tests the derive_crawl_params static method."""
    params = Crawler.derive_crawl_params(url)
    assert params == expected_params


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "mock_extract_return, expected_exception, limit_override",  # Update parameter name
    [
        # Original test cases, updated for new return structure
        ({"page_is_noindex": False, "page_is_nofollow": False, "urls_to_crawl": ["url1", "url2"], "skipped_urls": []}, None, None),
        ({"side_effect": HTTPError("Test HTTP Error")}, CrawlerValidationHTTPError, None),
        (
            {"page_is_noindex": False, "page_is_nofollow": False, "urls_to_crawl": [f"url{i}" for i in range(6)], "skipped_urls": []},
            CrawlerValidationTooManyURLsError,
            5,
        ),
        # New test case: noindex and nofollow
        (
            {"page_is_noindex": True, "page_is_nofollow": True, "urls_to_crawl": ["url1"], "skipped_urls": ["url2"]},
            CrawlerValidationNoIndexNofollowError,
            None,
        ),
        # Test case: too many URLs, ensuring skipped_urls are ignored for the count
        (
            {
                "page_is_noindex": False,
                "page_is_nofollow": False,
                "urls_to_crawl": [f"url{i}" for i in range(6)],
                "skipped_urls": ["skipped1", "skipped2"],
            },
            CrawlerValidationTooManyURLsError,
            5,
        ),
    ],
    ids=[
        "success",
        "http_error",
        "too_many_urls_to_crawl",  # Update ID
        "noindex_and_nofollow",  # New ID
        "too_many_urls_excluding_skipped",  # New ID
    ],
)
@patch("es_knowledge_base_mcp.clients.crawl.extract_urls_from_webpage", new_callable=AsyncMock)
async def test_validate_crawl(mock_extract_urls, mock_extract_return, expected_exception, limit_override):  # Update parameter name
    """Tests validate_crawl scenarios using parametrization."""
    test_url = "http://example.com"  # Inlined value
    max_limit = 10  # Inlined value

    # Configure the mock based on the provided return value or side effect
    if "side_effect" in mock_extract_return:
        mock_extract_urls.side_effect = mock_extract_return["side_effect"]
    else:
        mock_extract_urls.return_value = mock_extract_return

    current_limit = limit_override if limit_override is not None else max_limit

    expected_derived_params = Crawler.derive_crawl_params(test_url)

    if expected_exception:
        with pytest.raises(expected_exception):
            await Crawler.validate_crawl(test_url, current_limit)
    else:
        params = await Crawler.validate_crawl(test_url, current_limit)
        assert params == expected_derived_params

    mock_extract_urls.assert_called_once_with(
        url=test_url,
        domain_filter=expected_derived_params["domain"],
        path_filter=expected_derived_params["filter_pattern"],
    )


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "domain, seed_url, filter_pattern, index_name, es_settings",
    [
        ("http://example.com", "http://example.com/start", "/start.*", "test_index", {"host": "localhost", "port": 9200}),
        ("http://example.com", "http://example.com/start", "", "test_index", {"host": "localhost", "port": 9200}),
        ("http://example.com", "http://example.com/start", "/start.*", "test_index", {}),
        (
            "https://sub.example.org",
            "https://sub.example.org/docs/page1",
            "/docs/.*",
            "docs_index",
            {"host": "es-host", "port": 9201, "user": "elastic"},
        ),
        ("http://minimal.com", "http://minimal.com/", "/", "minimal_idx", {"host": "127.0.0.1"}),
        ("https://secure.com", "https://secure.com/app", "/app.*", "secure_idx", {"host": "secure-es", "port": 9200, "use_ssl": True}),
    ],
    ids=[
        "basic",
        "empty_filter",
        "empty_es_settings",
        "different_values",
        "minimal_es",
        "https_ssl",
    ],
)
async def test__prepare_crawl_config_file(
    snapshot,
    domain: str,
    seed_url: str,
    filter_pattern: str,
    index_name: str,
    es_settings: dict,
):
    """Tests _prepare_crawl_config_file using parametrization and snapshot testing."""
    inject_file = await Crawler._prepare_crawl_config_file(domain, seed_url, filter_pattern, index_name, es_settings)

    assert isinstance(inject_file, InjectFile)
    assert inject_file == snapshot
