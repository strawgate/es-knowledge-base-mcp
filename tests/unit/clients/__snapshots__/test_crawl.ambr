# serializer version: 1
# name: test__prepare_crawl_config_file[basic]
  InjectFile(filename='/config/crawl.yml', content='domains:\n- crawl_rules:\n  - pattern: /start.*\n    policy: allow\n    type: begins\n  - pattern: .*\n    policy: deny\n    type: regex\n  seed_urls:\n  - http://example.com/start\n  url: http://example.com\nelasticsearch:\n  host: localhost\n  port: 9200\nlog_level: DEBUG\noutput_index: test_index\noutput_sink: elasticsearch\n')
# ---
# name: test__prepare_crawl_config_file[different_values]
  InjectFile(filename='/config/crawl.yml', content='domains:\n- crawl_rules:\n  - pattern: /docs/.*\n    policy: allow\n    type: begins\n  - pattern: .*\n    policy: deny\n    type: regex\n  seed_urls:\n  - https://sub.example.org/docs/page1\n  url: https://sub.example.org\nelasticsearch:\n  host: es-host\n  port: 9201\n  user: elastic\nlog_level: DEBUG\noutput_index: docs_index\noutput_sink: elasticsearch\n')
# ---
# name: test__prepare_crawl_config_file[empty_es_settings]
  InjectFile(filename='/config/crawl.yml', content='domains:\n- crawl_rules:\n  - pattern: /start.*\n    policy: allow\n    type: begins\n  - pattern: .*\n    policy: deny\n    type: regex\n  seed_urls:\n  - http://example.com/start\n  url: http://example.com\nelasticsearch: {}\nlog_level: DEBUG\noutput_index: test_index\noutput_sink: elasticsearch\n')
# ---
# name: test__prepare_crawl_config_file[empty_filter]
  InjectFile(filename='/config/crawl.yml', content="domains:\n- crawl_rules:\n  - pattern: ''\n    policy: allow\n    type: begins\n  - pattern: .*\n    policy: deny\n    type: regex\n  seed_urls:\n  - http://example.com/start\n  url: http://example.com\nelasticsearch:\n  host: localhost\n  port: 9200\nlog_level: DEBUG\noutput_index: test_index\noutput_sink: elasticsearch\n")
# ---
# name: test__prepare_crawl_config_file[https_ssl]
  InjectFile(filename='/config/crawl.yml', content='domains:\n- crawl_rules:\n  - pattern: /app.*\n    policy: allow\n    type: begins\n  - pattern: .*\n    policy: deny\n    type: regex\n  seed_urls:\n  - https://secure.com/app\n  url: https://secure.com\nelasticsearch:\n  host: secure-es\n  port: 9200\n  use_ssl: true\nlog_level: DEBUG\noutput_index: secure_idx\noutput_sink: elasticsearch\n')
# ---
# name: test__prepare_crawl_config_file[minimal_es]
  InjectFile(filename='/config/crawl.yml', content='domains:\n- crawl_rules:\n  - pattern: /\n    policy: allow\n    type: begins\n  - pattern: .*\n    policy: deny\n    type: regex\n  seed_urls:\n  - http://minimal.com/\n  url: http://minimal.com\nelasticsearch:\n  host: 127.0.0.1\nlog_level: DEBUG\noutput_index: minimal_idx\noutput_sink: elasticsearch\n')
# ---
